### DISCUSSION TITLE
Live data with websocket

### DISCUSSION THREAD

**Discussion by tomtom94** [TanStack/query#1519] [Q&A]
Hello,

I noticed you can't access the reducers, middlewares, global state etc... I just noticed you can't access anything ;)

What if I wanna make a live chat with websocket.

I receive a push from the server containing the next message of the conversation, 
I wanna add this message to the top of the array messages downloaded previously with your `useQuery` method. 

How can I do this just give me a link from your doc.

I hope for you this is possible. If not possible, would you give us finally access to the reducers ? exactly like redux does.

Thanks.

**Comment by tannerlinsley** [comment]
A websocket message can be thought of as the result of a mutation, so the same APIs here should be used. 

https://react-query.tanstack.com/guides/updates-from-mutation-responses

**Comment by tomtom94** [comment]
Ok nice exactly like Redux I am reassured.

I have learned all the Redux in the market currently (thunk, saga, observable) but I don't see the point of `react-query` seriously... 

To me this is natural every developer knows the main state of a request `isLoading` `isError` you don't need a framework for that.

Another question how can I catch the http request code response in case of error, I wanna know if this is a code 404 or 403. How can I save that beside the error message.

Have a look how I am doing this for clients https://github.com/tomtom94/react-easy-ssr/blob/master/src/store/reducers/index.ts#L39

Would you like to convince me ? or you don't care at all and you understand my point of view. Just need to know.

**Comment by tannerlinsley** [comment]
> I have learned all the Redux in the market currently (thunk, saga, observable)

React Query has nothing to do with Redux, so I'm not sure why this is relevant here.

> I don't see the point of react-query seriously...

If you have read the documentation, watched the videos, tried it out, etc and still don’t see the use case or benefit then unfortunately there’s little left for me to do to help you understand or convince yourself. No one is pressuring you to use it.

> every developer knows the main state of a request `isLoading/isError`

The state representation for declarative data fetching, stale-while-revalidate and background data fetching is no cakewalk. Conflating the concept to 2 booleans is exactly why Redux data fetching has been stuck in a rut for years.

> you don't need a framework for that

Framework? Nope. But if your application is meant to (at the very least) "appear" to be in sync with your server-side data, then you absolutely need something to standardize those contracts. Of course, you can always attempt to manually replicate all of the functionality of React Query in your own custom Redux setup. I spent 3 years doing just that and didn't really make a dent in the problem space, but if that's what it's going to take to convince yourself, then I say go for it. Your learning experience will be priceless. Or, for the sake of time, you could perhaps entertain the thought that many others have also “learned all the Redux the market” and gone on to create better tools that can save you from further hassle. I've never seen any tool in the Redux ecosystem come close to the UX/DX, features, and efficiency of React Query. Even RTK-Query (at the time of writing) is extremely nascent with a long road ahead.

> Another question how can I catch the http request code response in case of error, I wanna know if this is a code 404 or 403. How can I save that beside the error message.
> Have a look how I am doing this for clients https://github.com/tomtom94/react-easy-ssr/blob/master/src/store/reducers/index.ts#L39

It appears from this comment that you still didn't bother to read the docs. I could be mistaken, so [you can read about throwing errors in your query functions here](https://react-query.tanstack.com/guides/queries#query-basics)

> Would you like to convince me?

Sure, but I haven't the time.

> or you don't care at all

Untrue. I try to treat everyone's questions with a minimum amount of respect and care, as long as they do the same for me.

> and you understand my point of view

I understand that it can be difficult to learn new data fetching concepts if you've been entrenched in the Redux ecosystem for too long.

**Comment by tomtom94** [comment]
Yes sure,

I read the doc very quickly, but it is said `React Query is a server-state library` but I just realised by myself I can use `useQuery` in an onClick event on the client side to manage the famous `client-state` ;)

<img width="1167" alt="Capture d’écran 2020-12-24 à 15 43 50" src="https://user-images.githubusercontent.com/262550/103095144-dce67580-45ff-11eb-8d75-94474b77d7dd.png">

I also notice you allow the server side rendering via Next.js with the method `getServerSideProps`. Then it means you are finally using the global variable via Next.js (Writing in the `window` is the only one way to pass data from server to client)

Your module is really great. Congrats to you and many thanks for your help.

**Reply by redbar0n** [reply]
Don't confuse server-state with server-side-state. 

* React Query is a library for handling server-state (a recently coined concept) on the client side. Meaning: state that should be persisted from the client to the server.

* Whereas Redux, MobX, Zustand etc. are libraries that _should_ be used to handle client-only-state on the client side. But some people use them to _also_ handle server-state on the client side too, and run into a heap of problems, which React Query would otherwise have solved for them. Like syncing and caching, like @tannerlinsley mentioned.

That page is confusing, I know. I struggled with this too.
